ddtrace/internal/processor.py

```diff
@@ -12,18 +15,80 @@
 
 
 @attr.s
-class TraceProcessor(object):
-    _filters = attr.ib(type=List[TraceFilter])
+class Processor(six.with_metaclass(abc.ABCMeta)):
+    """A Processor is used to process spans as they are created and finished by a tracer."""
+
+    def __attrs_post_init__(self):
+        # type () -> None
+        """Default post initializer which logs the representation of the
+        Processor at the ``logging.DEBUG`` level.
+
+        The representation can be modified with the ``repr`` argument to the
+        attrs attribute::
+
+            @attr.s
+            class MyProcessor(Processor):
+                field_to_include = attr.ib(repr=True)
+                field_to_exclude = attr.ib(repr=False)
+        """
+        log.debug("initialized processor %r", self)
+
+    @abc.abstractmethod
+    def on_span_start(self, span):
+        # type: (Span) -> None
+        """Called when a span is started.
+
+        This method is useful for making upfront decisions on spans.
+
+        For example, a sampling decision can be made when the span is created
+        based on its resource name.
+        """
+        pass
+
+    @abc.abstractmethod
+    def on_span_finish(self, data):
+        # type: (Any) -> Any
```

Comment:
Not sure about this `Any -> Any` thing, it's really hard to write anything meaningful with `Any`. Is there any way to have a proper type there?

Reply:
Yeah it's tricky to type it for the abstract processor since processors can define their own types between each other (unless we want to add some structure here).

I think it's fine because implementors of processor can define their own stricter types, see #2038 for how that works between each of the processors as an example

Reply:
The problem that I see is that it's impossible to contruct a Processor without finally doing a bunch of `isinstance`-like calls or being sure in which order they are plugged together.

I see in #2038 most are using a `Span` or a `List[Span]`. Couldn't that be a common type for all processors to work on?

Reply:
Yeah the ordering is important but I think that makes sense - unfortunately we can't get static type checking on it because the processor chain is created at run-time. We could possibly verify it ourselves when constructing the chain -although I'm not sure if we can access the type hints from comments-style typing.

Even without verification I think it makes sense that if I have a processor that only accepts traces (`List[Span]`) then it needs to follow the processor which collects spans into a trace. Similarly if I had a processor that only acted on spans.

Reply:
I think I'm still confused about what else than a `List[Span]` could be passed to a `Processor`?

Reply:
Right now really we only deal with lists of spans but in the future it's possible that we might only process single spans through the chain. Plus the first processor has to accept a single span. We could type it as `Union[Span, List[Span]]` since those are the possibilities right now (and foreseeably)

Reply:
We probably need to keep it as `List[Span]` forever, otherwise a bunch of custom filters people have implemented are going to break.

```python
def filter_healthchecks(trace):
    root = trace[0]
    if root.get_tag("http.url") == "/healthcheck":
        return None
    return trace
```

Having a change where the filter is not called on a list of spans will be a breaking change.

Reply:
Yeah trace filters are going to be incompatible with span streaming (without defeating the purpose of span streaming)

Reply:
should be able to keep them as long as "stitcher" exists, but in order to remove stitcher, we'll need to do a breaking change with migration docs, etc

Reply:
Oh to clarify I mean span streaming to the agent (we wouldn't need to stitch the spans together into a trace).

Reply:
> Right now really we only deal with lists of spans but in the future it's possible that we might only process single spans through the chain. Plus the first processor has to accept a single span. We could type it as Union[Span, List[Span]] since those are the possibilities right now (and foreseeably)

Well, if the future needs `Span` then you'll need something different… a `SpanProcessor` if you want, and a different mechanism. Doing a one-size-fits-all design is not going to help us maintain a proper, robust and optimized system. 😞 

Am I right if I understand you're trying to design this to also work for span streaming, where we could process on a per-span rather than a per-trace basis?

If span streaming is on, any filter/processor based on trace won't work anymore? If I have a filter that drops every trace made of more than 200 spans, how would I do that with span streaming?

Reply:
> Well, if the future needs Span then you'll need something different… a SpanProcessor if you want, and a different mechanism. Doing a one-size-fits-all design is not going to help us maintain a proper, robust and optimized system. 😞

Yeah, no I'm not a fan of trying to guess what the future may hold. I think really the problem right now is that processors can be chained together _and_ the processors may change the format of the data returned. Trace filters break the processor model because the type of data it consumes is dependent on a processor.

I think the right move would be that trace filters become part of the `SpansToTraceProcessor` (#2038). Processors then only act on spans and no state is passed between processors. The only problem with this is that the current writer wouldn't make sense - it would have to be owned by a processor/be a processor. This is roughly the model that opentelemetry has.

> Am I right if I understand you're trying to design this to also work for span streaming, where we could process on a per-span rather than a per-trace basis?

Yes, ideally I think we'd only want to do on a per-span basis.

> If span streaming is on, any filter/processor based on trace won't work anymore? If I have a filter that drops every trace made of more than 200 spans, how would I do that with span streaming?

Correct, trace filters won't work in a span streaming case. The trade-off would be that you don't have 200+ span traces kept in memory. I'm not sure if trace filters are something we want to support long term as right now they don't work with async/partial flushing due to traces being split up (this is fixed for async in #2038).